src/parallel.cpp

```diff
@@ -549,6 +549,10 @@ void HandleBlockMessageThread(CNode *pfrom, const string strCommand, const CBloc
             if (nDoS > 0)
                 dosMan.Misbehaving(pfrom, nDoS);
         }
+
+        // the current fork is bad due to this block so reset the best header to the best fully-validated block
+        // so we can download another fork of headers (and blocks).
+        pindexBestHeader = chainActive.Tip();
```

Comment:
i don't see anything wrong with what you did, setting pindexBestHeader to chainActive.Tip() is fine...but what's missing is that because we're multithreaded, while we were processing the block we could have received more headers both on the good and possibly the bad chain...The bad chain headers are not of interest because as soon as you invalidate the block those headers will not be considered when you do FindMostWorkChain()....however, if you received more headers on the good chain, now what you're doing is setting pindexBestHeader to be less than what it actually is...so that's why you need to do what is below and only set it to be equal to chainActive if no other headers have arrived:

if (FindMostWorkChain()->nChainWork > chainActive.Tip()->nChainWork)
{
pindexBestHeader = FindMostWorkChain();
}
else
{
pindexBestHeader = chainActive.Tip();
}